### PR TITLE
Feature/COP-8885 Update Product Review Date

### PR DESCRIPTION
**Pull request by hiteshprayosha**
- Created endpoint to update product review date
- Update Camunda Job Timer
- Update existing S3 Object

### CODE REVIEW

File: src/main/java/io/digital/patterns/workflow/cases/CaseApiController.java [hunk 1]
```diff
@@ -65,5 +66,23 @@
...
+		try {
+			casesApplicationService.setSubmissionData(formBody, businessKey,
+					key, new PlatformUser(
+							(JwtAuthenticationToken) SecurityContextHolder.getContext().getAuthentication()
+					));
+		} catch (Exception e) {
+			log.error("Failed to process the put submission data request {}", e.getMessage());
+			return ResponseEntity.unprocessableEntity().body(e.getMessage());
+		}
+		return ResponseEntity.ok().build();
```
Comment (michaelruocco) [inline src/main/java/io/digital/patterns/workflow/cases/CaseApiController.java:85]: You could make this a bit neater and reduce the nesting a little by using some private methods here I think, e.g.

```
public ResponseEntity<Object> putSubmissionData(@PathVariable String businessKey,
	                                                @RequestParam String key,
	                                                @RequestBody String formBody) {
    try {
        log.debug("putSubmissionData businessKey {} and key {}", businessKey, key);
	casesApplicationService.setSubmissionData(formBody, businessKey, key, getUser());
	return ResponseEntity.ok().build();
    } catch (Exception e) {
        //its normally a good idea with unexpected errors to log the stacktrace too, you can do this by adding the exception as the last argument to the log message e.g.
        log.error("Failed to process the put submission data request {}", e.getMessage(), e);
        return ResponseEntity.unprocessableEntity().body(e.getMessage());
    }
}

private static PlatformUser getUser() {
    JwtAuthenticationToken authToken = (JwtAuthenticationToken)  SecurityContextHolder.getContext().getAuthentication();
    return new PlatformUser(authToken);			
}
```

File: src/main/java/io/digital/patterns/workflow/cases/CaseApiController.java [hunk 2]
```diff
@@ -65,5 +66,23 @@
         return ResponseEntity.ok(submissionData.toString());
     }
 
+	@PutMapping(path = "/{businessKey}/submission")
+	@ConditionalOnProperty(prefix = "update.timer", name = "enabled")
```
[the remark this answers is not in the document]
Comment (hiteshprayosha) [reply src/main/java/io/digital/patterns/workflow/cases/CaseApiController.java:70]: application.properties

File: src/main/java/io/digital/patterns/workflow/cases/CaseApiController.java (hunk 1 above)
Comment (hiteshprayosha) [reply src/main/java/io/digital/patterns/workflow/cases/CaseApiController.java]: Done, Good point

File: src/main/resources/application.yml [hunk 3]
```diff
@@ -1,3 +1,5 @@
+update:
+  timer.enabled: false
 keycloak:
```
[the remark this answers is not in the document]
Comment (michaelruocco) [reply src/main/resources/application.yml:3]: my bad, apologies I missed this from the bottom of your PR

File: src/main/java/io/digital/patterns/workflow/cases/CaseApiController.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (michaelruocco) [reply src/main/java/io/digital/patterns/workflow/cases/CaseApiController.java:66]: @hiteshprayosha has pointed out the config that I missed which is added here: src/main/resources/application.yml

File: src/main/java/io/digital/patterns/workflow/cases/CasesApplicationService.java [hunk 4]
```diff
@@ -366,4 +425,95 @@ private ObjectListing getObjectListingForBusinessKey(String businessKey) {
...
+
+		} catch (IOException | AmazonServiceException e) {
+			log.error("Failed to upload S3 form data for {}", formName);
+			throw new RuntimeException(e);
+		} finally {
+			if (scratchFile != null && scratchFile.exists()) {
+				scratchFile.delete();
+			}
+		}
+	}
```
[the remark this answers is not in the document]
Comment (hiteshprayosha) [reply src/main/java/io/digital/patterns/workflow/cases/CasesApplicationService.java:466]: S3Uploder is helping to upload form, here little different way to upload form object to S3. so we can not use this S3Uploder.

[the remark this answers is not in the document]
Comment (michaelruocco) [reply src/main/java/io/digital/patterns/workflow/cases/CasesApplicationService.java:466]: How is it different exactly? the code for this method is:

```
private String updateS3Object(String formName, String form, String submissionDataKey,
	                              ObjectMetadata latestMetadata) {
		log.info("Start to update s3 object");
		File scratchFile = null;
		try {
			//Create json file object
			scratchFile = File.createTempFile(UUID.randomUUID().toString(), ".json");

			FileUtils.copyInputStreamToFile(IOUtils.toInputStream(form, "UTF-8"), scratchFile);

			PutObjectRequest request = new PutObjectRequest(awsConfig.getCaseBucketName(), submissionDataKey,
					scratchFile);
			request.setMetadata(latestMetadata);

			final PutObjectResult putObjectResult = amazonS3Client.putObject(request);
			return submissionDataKey;

		} catch (IOException | AmazonServiceException e) {
			log.error("Failed to upload S3 form data for {}", formName);
			throw new RuntimeException(e);
		} finally {
			if (scratchFile != null && scratchFile.exists()) {
				scratchFile.delete();
			}
		}
	}
```

the code in the method I was referencing in the link above is:

```
            private boolean tryPerformUpload(String objectKey, String bucketName, ObjectMetadata metadata, Form form) {
        File tempFile = null;
        try {
            tempFile = writeToTemporaryFile(form);
            PutObjectRequest request = new PutObjectRequest(bucketName, objectKey, tempFile);
            request.setMetadata(metadata);
            log.info("uploading object with key {} to bucket {}", objectKey, bucketName);
            PutObjectResult result = amazonS3.putObject(request);
            log.info("Uploaded to S3 '{}'", result.getETag());
            return true;
        } catch (IOException | AmazonServiceException e) {
            throw new DataUploadFailedException(e);
        } finally {
            if (tempFile != null && tempFile.exists()) {
                tempFile.delete();
            }
        }
    }
```

the only thing that is different is the error log message you've added, the exception type thats thrown (although your method should be throwing something more specific than runtime exception ideally, and the fact that one message returns true if it completes, whereas yours just returns the object key string that you pass in, but if you look at what they actually do 95% of the code is identical, and the bits that have been changed don't really need to be as far as I can see as they dont change the functionality anyway.

File: src/main/java/io/digital/patterns/workflow/cases/CasesApplicationService.java [hunk 5]
```diff
@@ -30,44 +31,54 @@
...
 
+	private static final String isoDatePattern = "yyyy-MM-dd'T'HH:mm:ssXXX";
+	private static final SimpleDateFormat simpleDateFormat = new SimpleDateFormat(isoDatePattern);
     private final HistoryService historyService;
     private final AmazonS3 amazonS3Client;
     private final RestHighLevelClient elasticsearchClient;
     private final AwsProperties awsConfig;
     private final CaseActionService caseActionService;
     private final AuthorizationService authorizationService;
+	private final ManagementService managementService;
```
[the remark this answers is not in the document]
Comment (hiteshprayosha) [reply src/main/java/io/digital/patterns/workflow/cases/CasesApplicationService.java]: Fixed

File: src/main/java/io/digital/patterns/workflow/cases/CasesApplicationService.java [hunk 6]
```diff
@@ -366,4 +425,95 @@ private ObjectListing getObjectListingForBusinessKey(String businessKey) {
...
+					} else {
+						array.put(i, String.valueOf(aObj));
+					}
+				}
+			} else {
+				o.put(key, valueOf(json));
+			}
+		}
+		return o;
+	}
```
[the remark this answers is not in the document]
Comment (hiteshprayosha) [reply src/main/java/io/digital/patterns/workflow/cases/CasesApplicationService.java:450]: Refactor

File: src/main/java/io/digital/patterns/workflow/cases/CasesApplicationService.java (hunk 4 above)
[the remark this answers is not in the document]
Comment (hiteshprayosha) [reply src/main/java/io/digital/patterns/workflow/cases/CasesApplicationService.java:444]: refactor

**Review by michaelruocco — COMMENTED** [review]
LGTM

**Review by michaelruocco — DISMISSED** [review]
LGTM

**Review by michaelruocco — APPROVED** [review]
LGTM

**Pull request merged**